kernel/src/hil/kv_store.rs

```diff
@@ -0,0 +1,112 @@
+//! Interface for Key-Value (KV) Stores
+
+use crate::returncode::ReturnCode;
+
+#[derive(Clone, Copy, PartialEq)]
+pub enum Operation {
+    None,
+    Init,
+    GetKey,
+    AppendKey,
+    InvalidateKey,
+    GarbageCollect,
+}
+
+/// The type of keys, this should define the output size of the digest
+/// operations.
+pub trait KeyType: Eq + Copy + Clone + Sized + AsRef<[u8]> + AsMut<[u8]> {}
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a, K: KeyType> {
+    /// This callback is called when any operation is complete.
+    ///
+    /// `operation`: The operation that completed
+    /// `result`: Nothing on success, 'ReturnCode' on error
+    /// `key`: A key buffer if one exists
+    /// `value`: A value buffer if one exists
+    /// `ret_buf`: A ret_buf buffer if one exists
+    fn operation_complete(
```

Comment:
I'd think most other HILs would use separate methods in the client for each type of callback (each kind of operation that is completed). Take for instance the nonvolatile storage driver:
https://github.com/tock/tock/blob/2762b10452a883a1a5d5a28c0d0831f6aee2c69d/kernel/src/hil/nonvolatile_storage.rs#L21-L32

That would have an advantage in that you can more precisely specify what values are returned for a specific operation (instead of wrapping them all in an `Option`) and thus reduces the potential for runtime errors and the necessity for checks.

I think this might be fine too. Is there a specific reason you designed it this way? Maybe because of code bloat, since most HILs probably don't have that many different callbacks?

Reply:
I had this same question. The `Option<>` seems like extra work for the user that separate functions would make simpler.

Reply:
I had combined them to allow other operations such as init to have callabacks. I have split them out though.